File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.cpp
class Solution {
public:
    vector<vector<int>>result;
    void dfs(vector<int>& candidates, int target, int curr, int n, vector<int>path){
        if(target < 0){
            return;
        }

        if(target == 0){
            if(count(result.begin(),result.end(),path) == 0)  result.push_back(path);
            return;
        }

        if(curr >= n){
            return;
        }
        
        path.push_back(candidates[curr]);
        dfs(candidates,target-candidates[curr],curr+1,n,path);
        path.pop_back();
        
        dfs(candidates,target,curr+1,n,path);
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        result.clear();
        sort(candidates.begin(),candidates.end());
        int n = candidates.size();
        dfs(candidates,target,0,n,{});
        return result;
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        """

        """

        n = len(nums)
        nums.sort()
        result = []

        for i in range(n):
            if i > 0 and nums[i] == nums[i-1]:
                continue

            left,right = i+1,n-1

            while left < right:
                target = nums[i] + nums[left] + nums[right]
                if target == 0:
                    result.append([nums[i], nums[left], nums[right]])
                    left += 1

                    while left < right and nums[left] == nums[left-1]:
                        left += 1
                
                elif target > 0:
                    right -= 1
                
                else:
                    left += 1

        return result
